File: typescript/manipulate.ts
/**
 * 
 * @param text string of value example "aabb", "abcde"
 * @returns {string[]} string array of result
 */ 
export function manipulate(text: string) : string[] {
    // TODO : Start your code here
    if (text.length <= 1) {
        return [text];
    }

    const chars = text.split('');
    const result = new Set<string>(); // Using a set to avoid duplicates

    chars.forEach((char, idx) => {
        const remainingChars = [...chars.slice(0, idx), ...chars.slice(idx + 1)].join('');
        const permutations = manipulate(remainingChars);
        permutations.forEach(permutation => {
            result.add(char + permutation);
        });
    });

    return Array.from(result); // Convert the set back to an array
    // throw new Error("Not implemented");
}

File: golang/manipulate.go
package main

func Manipulate(text []string) []string {
	// TODO : start your code here
	if len(text) <= 1 {
		if len(text) == 0 {
			return []string{}
		}
		return []string{text[0]}
	}

	permMap := make(map[string]bool)
	var result []string

	for i, char := range text {
		remainingChars := append(append([]string{}, text[:i]...), text[i+1:]...)
		for _, permutation := range Manipulate(remainingChars) {
			combined := char + permutation
			if _, exists := permMap[combined]; !exists {
				result = append(result, combined)
				permMap[combined] = true
			}
		}
	}

	return result
}
